### DISCUSSION TITLE
Using CPython's Frame API?

### DISCUSSION THREAD

**Discussion by AriMKatz** [jax-ml/jax#11120] [Ideas]
[TorchDynamo](https://github.com/pytorch/torchdynamo) is a new compilation pathway  for Pytorch which leverages CPython's Frame API. This API is purpose built for alternative compilers and other things like that. 

I'm still getting to know Jax, but looks like there are different ways of dealing with control flow, each with their own tradeoffs. Could a Torch Dynamo like approach be helpful in improving the situation so for example you don't have to use  `jax.lax.while_loop`, or would you still need some way to express those invariants, unless the compiler can pick up on it?

**Comment by YouJiacheng** [comment]
Hmmm, IIUC, this API only serves as an alternative to tracer.

**Comment by npuichigo** [comment]
[TorchDynamo](https://github.com/pytorch/torchdynamo) now support dynamic shape in PyTorch 2.1. But jax jit needs static shape. Any insights on that?
